contracts/samples/SimpleAccountFactory.sol

```diff
@@ -26,27 +27,35 @@ contract SimpleAccountFactory {
      * This method returns an existing account address so that entryPoint.getSenderAddress() would work even after account creation
      */
     function createAccount(address owner,uint256 salt) public returns (SimpleAccount ret) {
-        address addr = getAddress(owner, salt);
-        uint codeSize = addr.code.length;
-        if (codeSize > 0) {
+
+        bytes memory proxyConstructor = getProxyConstructor(owner);
+        // must use assembly and not "new{salt}()", so it doesn't revert on failure.
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            ret := create2(0, add(proxyConstructor, 0x20), mload(proxyConstructor), salt)
```

Comment:
The [EIP-1014](https://eips.ethereum.org/EIPS/eip-1014) states that the `creation throws immediately` in case there is already a code at the destination address, why do you expect it to return `address(0)`?

Reply:
TL;DR: you're right. it should fail according to the spec, but all node implementations support it. I'm not hesitate to rely on it in our reference impl.

According to the spec, create2 should revert if target already exists.
but we have an [explicit test](https://github.com/eth-infinitism/account-abstraction/blob/SimpleAccount-factory-optim/test/simple-wallet.test.ts#L164) - that shows this method works...

What I did find out that it works - but consumes more gas - estimating "create" takes ~280Kgas, but estimating again after create consumes >500Kgas (and 2M gas on hardhat)
Gas doesn't really matter, as it is meant for being a helper for off-chain method.

I think that what happens on failure is that the GO implementation doesn't revert, but consume all gas - but create behaves like a "call", and set aside 1/64 of the gas.
If that 1/64 gas left is enough, the method may succeed.
So if the view method is called with enough (e.g. default 5M gas), then the 1/64 excess that is left is enough to complete the createSender method....

solidity compiler is also aware if this behavior
```
     let expr_19_address := create2(0, _2, sub(_3, _2), expr_18_salt)
     if iszero(expr_19_address) { revert_forward_1() }
```

Reply:
This was the part where i was concerned about, for example, if there is a smart contract that relies on the `entrypoint.getSenderAddress()`, and do something afterwards that makes it non-view, then won't it be a easy honeypot to make end user to burn the eth gas? Yes it can be avoided by checking the gas limit and preview amount of gas spent but metamask does not show gas limit by default.

Maybe some easy prevention will be making it clear calling the `entryPoint.getSenderAddress()` in other contract is risky.

eg)
```
address factory;
...
function claimYourTokenToAA(bytes calldata initCode) external {
    address wallet = entryPoint.getSenderAddress(abi.encodePacked(factory,initCode)); // this will cost around 580k as you said
    token.transfer(wallet, 1e18);
}
```

Reply:
using initCode in an application seems a bit contrived (considering it requires entire constructor parameters).
It is expected to be filled by wallet software.

Reply:
Still, I'm a bit hesitant about including this as-is, as it depends on gas-left (and probably a misinterpretation of the "geth" code of the spec - though this implementation became the new standard)